### PR TITLE
[SPARK-16192][SQL] Add type checks in CollectSet

### PR DESCRIPTION
**Pull request by maropu**
## What changes were proposed in this pull request?

`CollectSet` cannot have map-typed data because MapTypeData does not implement `equals`. 
So, this pr is to add type checks in `CheckAnalysis`.
## How was this patch tested?

Added tests to check failures when we found map-typed data in `CollectSet`.

### CODE REVIEW

**Comment by maropu** [comment]
@cloud-fan Could you check this?

**Comment by hvanhovell** [comment]
@maropu could you implement `checkInputDataTypes` and do the validation in that method?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala [hunk 1]
```diff
@@ -73,9 +73,12 @@ trait CheckAnalysis extends PredicateHelper {
               s"invalid cast from ${c.child.dataType.simpleString} to ${c.dataType.simpleString}")
 
           case g: Grouping =>
-            failAnalysis(s"grouping() can only be used with GroupingSets/Cube/Rollup")
+            failAnalysis("grouping() can only be used with GroupingSets/Cube/Rollup")
           case g: GroupingID =>
-            failAnalysis(s"grouping_id() can only be used with GroupingSets/Cube/Rollup")
+            failAnalysis("grouping_id() can only be used with GroupingSets/Cube/Rollup")
+
+          case c: CollectSet if c.child.dataType.isInstanceOf[MapType] =>
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: This should be recursive. The child could be a struct/array containing a map.

Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: oh, my bad. I'll fix this.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala [hunk 2]
```diff
@@ -73,9 +73,12 @@ trait CheckAnalysis extends PredicateHelper {
               s"invalid cast from ${c.child.dataType.simpleString} to ${c.dataType.simpleString}")
 
           case g: Grouping =>
-            failAnalysis(s"grouping() can only be used with GroupingSets/Cube/Rollup")
+            failAnalysis("grouping() can only be used with GroupingSets/Cube/Rollup")
           case g: GroupingID =>
-            failAnalysis(s"grouping_id() can only be used with GroupingSets/Cube/Rollup")
+            failAnalysis("grouping_id() can only be used with GroupingSets/Cube/Rollup")
+
+          case c: CollectSet if !c.child.dataType.existsRecursively(_.isInstanceOf[MapType]) =>
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: Please do not use `CheckAnalysis` for this; move this check into the `CollectSet` expression by overriding `Expression.checkInputDataTypes`.

Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: okay

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisErrorSuite.scala [hunk 3]
```diff
@@ -435,6 +435,23 @@ class AnalysisErrorSuite extends AnalysisTest {
         "another aggregate function." :: Nil)
   }
 
+  test("we should fail analysis when we find map type data in collect_set") {
```
Comment (hvanhovell) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisErrorSuite.scala]: Lets remove this test.

Comment (maropu) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisErrorSuite.scala]: okay

**Comment by hvanhovell** [comment]
LGTM - pending jenkins

**Comment by hvanhovell** [comment]
Merging to master/2.0. Thanks!

**Pull request closed without merging**